src/bin/pgcopydb/pgsql_timeline.c

```diff
@@ -0,0 +1,371 @@
+#include "catalog.h"
+
+static void parseIdentifySystemResult(void *ctx, PGresult *result);
+static void parseTimelineHistoryResult(void *ctx, PGresult *result);
+
+
+typedef struct IdentifySystemResult
+{
+	char sqlstate[6];
+	bool parsedOk;
+	IdentifySystem *system;
+} IdentifySystemResult;
+
+
+typedef struct TimelineHistoryResult
+{
+	char sqlstate[6];
+	bool parsedOk;
+	char filename[MAXPGPATH];
+	char *content;
+} TimelineHistoryResult;
+
+/*
+ * pgsql_identify_system connects to the given pgsql client and issue the
+ * replication command IDENTIFY_SYSTEM. The pgsql connection string should
+ * contain the 'replication=1' parameter.
+ */
+bool
+pgsql_identify_system(PGSQL *pgsql, IdentifySystem *system, void *timelineContext)
+{
+	bool connIsOurs = pgsql->connection == NULL;
+
+	PGconn *connection = pgsql_open_connection(pgsql);
+	if (connection == NULL)
+	{
+		/* error message was logged in pgsql_open_connection */
+		return false;
+	}
+
+	/* extended query protocol not supported in a replication connection */
+	PGresult *result = PQexec(connection, "IDENTIFY_SYSTEM");
+
+	if (!is_response_ok(result))
+	{
+		log_error("Failed to IDENTIFY_SYSTEM: %s", PQerrorMessage(connection));
+		PQclear(result);
+		clear_results(pgsql);
+
+		PQfinish(connection);
+
+		return false;
+	}
+
+	IdentifySystemResult isContext = { { 0 }, false, system };
+
+	(void) parseIdentifySystemResult((void *) &isContext, result);
+
+	PQclear(result);
+	clear_results(pgsql);
+
+	log_sql("IDENTIFY_SYSTEM: timeline %d, xlogpos %s, systemid %" PRIu64,
+			system->timeline,
+			system->xlogpos,
+			system->identifier);
+
+	if (!isContext.parsedOk)
+	{
+		log_error("Failed to get result from IDENTIFY_SYSTEM");
+		PQfinish(connection);
+		return false;
+	}
+
+	/* while at it, we also run the TIMELINE_HISTORY command */
+	if (system->timeline > 1)
+	{
+		TimelineHistoryResult hContext = { 0 };
+
+		char sql[BUFSIZE] = { 0 };
+		sformat(sql, sizeof(sql), "TIMELINE_HISTORY %d", system->timeline);
+
+		result = PQexec(connection, sql);
+
+		if (!is_response_ok(result))
+		{
+			log_error("Failed to request TIMELINE_HISTORY: %s",
+					  PQerrorMessage(connection));
+			PQclear(result);
+			clear_results(pgsql);
+
+			PQfinish(connection);
+
+			return false;
+		}
+
+		(void) parseTimelineHistoryResult((void *) &hContext, result);
+
+		PQclear(result);
+		clear_results(pgsql);
+
+		if (!hContext.parsedOk)
+		{
+			log_error("Failed to get result from TIMELINE_HISTORY");
+			PQfinish(connection);
+			return false;
+		}
+
+		if (!parseTimelineHistory(hContext.filename, hContext.content, system,
+								  timelineContext))
+		{
+			/* errors have already been logged */
+			PQfinish(connection);
+			return false;
+		}
+
+		TimelineHistoryEntry *current = &(system->currentTimeline);
+
+		log_sql("TIMELINE_HISTORY: \"%s\", timeline %d started at %X/%X",
+				hContext.filename,
+				current->tli,
+				(uint32_t) (current->begin >> 32),
+				(uint32_t) current->begin);
```

Comment:
Let's use LSN_FORMAT_ARGS here?

Reply:
Yeah I realize it was done that way before your PR. You just happen to have a good opportunity to fix this in passing, thanks for doing that!